### DISCUSSION TITLE
Equivalent to middleware.ts for running logic before sending a response?

### DISCUSSION THREAD

**Discussion by zecka** [vercel/next.js#72240] [Help]
### Summary

Hello everyone,

In Next.js, I know we can create a middleware.ts file to execute code on all incoming requests before the rest of the code is processed. However, I am wondering if there is an equivalent way to run logic right before Next.js sends the response back to the client.

My use case is to add some custom logic after the request has been processed but right before the response is sent. I want this to apply globally to all pages and API routes.

Does Next.js provide any built-in mechanism for this type of functionality, or is there an established pattern to achieve this? Any insights or examples would be greatly appreciated.

Thanks in advance!

**Comment by icyJoseph** [comment]
The next.js middleware allows you to play around with the response rather. However, it is, as the request arrives, that you generate a response object, in which you can add headers and such. Typically with `NextResponse.next()`.

**Reply by zecka** [reply]
Yes, but Middleware is executed before the page rendering. My goal is to execute code at the end of the chain. For example, to ensure that all pages have a "cache-control" header with a rule like: if cache-control is not defined, then add cache-control = max-age 60.
